tools/codegen/api/cpp.py

```diff
@@ -153,14 +153,23 @@ def returns_type(rs: Sequence[Return]) -> str:
     'None': 'c10::nullopt',  # UGH this one is type directed
     'Mean': 'at::Reduction::Mean',
     '[]': '{}',
-    '[0,1]': '{0,1}',  # TODO: stop special casing
     'contiguous_format': 'MemoryFormat::Contiguous',
 }
 
 # Convert a JIT default into C++ expression representing the default
 def default_expr(d: str, t: Type) -> str:
     if d == 'None' and str(t) == 'Tensor?':
         return '{}'
+
+    if isinstance(t, OptionalType):
+        if d == 'None':
+            return 'c10::nullopt'
+
+        return default_expr(d, t.elem)
+
+    if isinstance(t, ListType) and d.startswith('[') and d.endswith(']'):
+        return '{' + d[1:-1] + '}'
```

Comment:
This is needed to handle `dim=[-2,-1]` in #45164 and more generally, to support arbitrary list defaults. Without it, it tries to literally put square-brackets into the C++ defaults.

Reply:
I'd prefer it if we can make this fail noisily if someone accidentally, e.g., prefixes the default value with an extra space. Can we do something like, assert `d.startswith('[') and d.endswith(']')` when `isinstance(t, ListType)`? That would ensure that the person specified braces.

Reply:
That check was there to allow sized lists with scalar defaults e.g. `int[2] padding=0`. So, I've put in an exception for unsized lists only.